docs/design_proposals/sync-improvements.md

```diff
@@ -0,0 +1,265 @@
+# Sync improvements
+
+* Author(s): Cornelius Weig (@corneliusweig)
+* Design Shepherd: Techal Desai (@tejal29)
+* Date: 03/20/2019
+* Status: Draft
+
+## Background
+
+Currently, skaffold config supports `artifact.sync` as a way to sync files
+directly to pods. So far, artifact sync requires a specification of sync
+patterns like
+
+```yaml
+sync:
+  '*.js': app/
+```
+
+This is error prone and unnecessarily hard to use, because the destination is
+already contained in the Dockerfile for docker build. (see #1166, #1581).
+In addition, the syncing needs to handle special cases for globbing and often
+requires a long list of sync patterns (#1807).
+
+Furthermore, builders should be able to inform Skaffold about additional sync paths (#1704).
+This will result in much faster deploy cycles, without users having to bother about these optimizations.
+
+## Design
+
+### Goals
+The sync mechanism in skaffold should be able to infer the destination location of local files.
+
+For example: given the following Dockerfile
+
+```dockerfile
+FROM nginx
+WORKDIR /var
+COPY static *.html www/
+ADD nginx.conf .
+```
+
+The sync locations should be inferred as
+- `nginx.conf` -> `/var/nginx.conf`
+- `static/page1.html` -> `/var/www/page1.html` (! no static folder at destination !)
+- `static/assets/style.css` -> `/var/www/assets/style.css`
+- `index.html` -> `/var/www/index.html`
+
+A fundamental change compared to the current sync mechanism is that one source path may by sync'ed to multiple destinations in the future.
+
+### Assumptions
+
+Builders know where to put static files in the container.
+This information needs to be pulled from the builders and made available in Skaffold.
+Currently, we know how to do this for Docker artifacts (@r2d4 suggested this in #1166).
+Whether jib and bazel builders can provide that information too, is unclear at the moment.
+
+### Interface changes
+
+There are at least two relevant changes
+
+1. When builders are queried for their dependencies, they must provide a list of destinations together with local source paths. The signature of `Builder.GetDependencies` (pkg/skaffold/build/build.go) therefore needs to change to:
+   ```go
+   type Builder interface {
+       Labels() map[string]string
+
+       Build(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact) ([]Artifact, error)
+
+       // OLD
+       // DependenciesForArtifact(ctx context.Context, artifact *latest.Artifact) ([]string, error)
```

Comment:
Was `[]string` before, just a list of local destination paths?

Reply:
Yes, only the local paths were extracted, thus `[]string`. What we want is now a mapping of local paths to several remote destinations.